Source/Core/Core/IOS/Network/WD/Command.cpp

```diff
@@ -4,33 +4,333 @@
+{
+  if (m_ipc_owner_fd < 0)
+  {
+    const auto flags = u32(request.flags);
+    const auto mode = WD::Mode(flags & 0xFFFF);
+    const auto buffer_flags = flags & 0x7FFF0000;
+    INFO_LOG_FMT(IOS_NET, "Opening with mode={} buffer_flags={:08x}", mode, buffer_flags);
+
+    // We don't support anything other than mode 1 and mode 3 at the moment.
+    if (mode != WD::Mode::DSCommunications && mode != WD::Mode::AOSSAccessPointScan)
+    {
+      ERROR_LOG_FMT(IOS_NET, "Unsupported WD operating mode: {}", mode);
+      DolphinAnalytics::Instance().ReportGameQuirk(GameQuirk::USES_UNCOMMON_WD_MODE);
+      return GetDefaultReply(s32(ResultCode::UnavailableCommand));
+    }
+
+    if (m_target_status == Status::Idle && mode <= WD::Mode::Unknown6)
+    {
+      m_mode = mode;
+      m_ipc_owner_fd = request.fd;
+      m_buffer_flags = buffer_flags;
+    }
+  }
+
+  INFO_LOG_FMT(IOS_NET, "Opened");
+  return Device::Open(request);
+}
+
+IPCCommandResult NetWDCommand::Close(u32 fd)
 {
-  s32 return_value = IPC_SUCCESS;
+  if (m_ipc_owner_fd < 0 || fd != u32(m_ipc_owner_fd))
+  {
+    ERROR_LOG_FMT(IOS_NET, "Invalid close attempt.");
+    return GetDefaultReply(u32(ResultCode::InvalidFd));
+  }
+
+  INFO_LOG_FMT(IOS_NET, "Closing and resetting status to Idle");
+  m_target_status = m_status = Status::Idle;
 
+  m_ipc_owner_fd = -1;
+  m_wake_up_proxy_request_thread.Set();
+  return Device::Close(fd);
+}
+
+IPCCommandResult NetWDCommand::SetLinkState(const IOCtlVRequest& request)
+{
+  const auto* vector = request.GetVector(0);
+  if (!vector)
+    return GetDefaultReply(u32(ResultCode::IllegalParameter));
+
+  const u32 state = Memory::Read_U32(vector->address);
+  INFO_LOG_FMT(IOS_NET, "WD_SetLinkState called (state={}, mode={})", state, m_mode);
+
+  if (state == 0)
+  {
+    if (!WD::IsValidMode(m_mode))
+      return GetDefaultReply(u32(ResultCode::UnavailableCommand));
+
+    INFO_LOG_FMT(IOS_NET, "WD_SetLinkState: setting target status to 1 (Idle)");
+    m_target_status = Status::Idle;
+  }
+  else
+  {
+    if (state != 1)
+      return GetDefaultReply(u32(ResultCode::IllegalParameter));
+
+    if (!WD::IsValidMode(m_mode))
+      return GetDefaultReply(u32(ResultCode::UnavailableCommand));
+
+    const auto target_status = GetTargetStatusForMode(m_mode);
+    if (m_status != target_status && m_info.enabled_channels == 0)
+      return GetDefaultReply(u32(ResultCode::UnavailableCommand));
+
+    INFO_LOG_FMT(IOS_NET, "WD_SetLinkState: setting target status to {}", target_status);
+    m_target_status = target_status;
+  }
+
+  return GetDefaultReply(IPC_SUCCESS);
+}
+
+IPCCommandResult NetWDCommand::GetLinkState(const IOCtlVRequest& request) const
+{
+  INFO_LOG_FMT(IOS_NET, "WD_GetLinkState called (status={}, mode={})", m_status, m_mode);
+  if (!WD::IsValidMode(m_mode))
+    return GetDefaultReply(u32(ResultCode::UnavailableCommand));
+
+  return GetDefaultReply(m_status == GetTargetStatusForMode(m_mode));
```

Comment:
What kind of return values are expected here? The error path returns a result code, but this success path returns a boolean.

Reply:
Yes, it's supposed to return a boolean here (converted to an integer). I'll make this more explicit with an int cast.